src/Microsoft.AspNetCore.Mvc.ApiExplorer/DefaultApiDescriptionProvider.cs

```diff
@@ -387,25 +396,46 @@ private string GetRelativePath(RouteTemplate parsedTemplate)
                 contentTypes.Add((string)null);
             }
 
-            foreach (var contentType in contentTypes)
+            var responseFormatMetadataProviders = _outputFormatters.OfType<IApiResponseFormatMetadataProvider>();
+
+            foreach (var responseType in responseTypes)
             {
-                foreach (var formatter in _outputFormatters)
+                if (responseType.Value == typeof(void))
                 {
-                    var responseFormatMetadataProvider = formatter as IApiResponseFormatMetadataProvider;
-                    if (responseFormatMetadataProvider != null)
+                    results.Add(new ApiResponseFormat()
                     {
-                        var supportedTypes = responseFormatMetadataProvider.GetSupportedContentTypes(contentType, type);
+                        StatusCode = responseType.Key,
+                        ResponseType = responseType.Value
+                    });
 
-                        if (supportedTypes != null)
+                    continue;
+                }
+
+                var responseTypeModelMetadata = _modelMetadataProvider.GetMetadataForType(responseType.Value);
+
+                foreach (var contentType in contentTypes)
+                {
+                    foreach (var responseFormatMetadataProvider in responseFormatMetadataProviders)
+                    {
+                        var formatterSupportedContentTypes = responseFormatMetadataProvider.GetSupportedContentTypes(
+                            contentType,
+                            responseType.Value);
+
+                        if (formatterSupportedContentTypes == null)
                         {
-                            foreach (var supportedType in supportedTypes)
+                            continue;
+                        }
+
+                        foreach (var formatterSupportedContentType in formatterSupportedContentTypes)
+                        {
+                            results.Add(new ApiResponseFormat()
                             {
-                                results.Add(new ApiResponseFormat()
-                                {
-                                    Formatter = formatter,
-                                    MediaType = supportedType,
-                                });
-                            }
+                                Formatter = (IOutputFormatter)responseFormatMetadataProvider,
+                                MediaType = formatterSupportedContentType,
+                                ResponseType = responseType.Value,
+                                StatusCode = responseType.Key,
+                                ResponseModelMetadata = responseTypeModelMetadata
+                            });
```

Comment:
I like this idea but probably there is one scenario where it might not work well. Each formatter looks at the type also to decide if they can support to write or not, so we having a list of types against a formatter would not be valid in this case.